src/Arcus.Security.Tests.Unit/Core/CachedSecretProviderTests.cs

```diff
@@ -73,73 +72,171 @@ public void CachedSecretProvider_CreateWithCorrectArguments_ShouldSucceed()
             Assert.NotNull(cachedSecretProvider);
         }
 
-        [Fact]
-        public async Task CachedSecretProvider_TwoCallsWithinCacheInterval_ShouldGetSameValueTwice()
+        [Theory]
+        [InlineData(GetCachedSecret.WithinCacheInterval)]
+        [InlineData(GetCachedSecret.OutsideCacheInterval)]
+        [InlineData(GetCachedSecret.SkippedCache)]
+        public async Task CachedSecretProvider_GetTwoSecretValues_WithCaching(GetCachedSecret cacheSetup)
         {
             // Arrange
-            string secretKeyValue = Guid.NewGuid().ToString("N");
-            var testSecretProvider = new TestSecretProviderStub(secretKeyValue);
-            var memCache = new MemoryCache(new MemoryCacheOptions());
-            var cacheConfiguration = new CacheConfiguration(TimeSpan.FromSeconds(3));
-            string keyName = "MyValue";
+            string expectedFirstSecret = Guid.NewGuid().ToString("N");
+            string expectedSecondSecret = Guid.NewGuid().ToString("N");
 
             // Act 
-            ICachedSecretProvider cachedSecretProvider = new CachedSecretProvider(testSecretProvider, cacheConfiguration, memCache);
-            var firstValue = await cachedSecretProvider.Get(keyName);
-            testSecretProvider.SecretValue = Guid.NewGuid().ToString("N"); // Change actual value on the internal secret provider !
-            var secondValue = await cachedSecretProvider.Get(keyName);
+            (string actualFirstSecret, string actualSecondSecret) = await GetTwoSecretsFromCachedProvider(
+                (expectedFirstSecret, expectedSecondSecret),
+                (provider, keyName) => provider.Get(keyName),
+                cacheSetup);
 
             // Assert
-            Assert.Equal(firstValue, secondValue);
-            Assert.Equal(1, testSecretProvider.CallsMadeSinceCreation);
+            Assert.True(
```

Comment:
If you need to add checks like `cacheSetup == GetCachedSecret.WithinCacheInterval` is inline data the best approach? I'd argue that this means that it should have a dedicated test case.

Certainly if the second assert starts with `cacheSetup == GetCachedSecret.OutsideCacheInterval ` and the 3rd scenario will thus not have a assertion.

Reply:
:) this approach is used quite a lot in property-based testing because of it generic usability. By abstracting the setup in readable values (enums) we quickly spot that this test file contains three different "properties" that will test three different system-endpoints.

The assertion you describe, is actually the idea behind "test-properties" that you define a relationship between test-input and test-output.

`cacheSetup == WithinCacheInterval == (actual1st == actual2st && expect1st == actual1st)`
becomes "_when the secret is retrieved within the cache interval, the same secret will be returned_".
Which is what we in fact want to test.

Reply:
I certainly get why you are taking this approach and appreciate the suggestion, but I'm afraid this will decrease the accessibility for new folks on the project.

Personally, I think it's better to stick with the current approach where we create the objects in every test, do the logic and have simple asserts.

If the setup is different, and thus requires different asserting, then that means we need a separate test case.

Would it be possible to make the changes accordingly, please?

Reply:
Ok, I look into this.

Reply:
Thanks!

Reply:
Ok, this is done I think?